src/crimson/net/SocketConnection.cc

```diff
@@ -759,95 +774,155 @@ seastar::future<> SocketConnection::repeat_connect()
   ceph_assert(!socket);
   peer_addr = _peer_addr;
   peer_type = _peer_type;
   messenger.register_conn(this);
   state = state_t::connecting;
-  return seastar::connect(peer_addr.in4_addr())
-    .then([this](seastar::connected_socket fd) {
-      if (state == state_t::closing) {
-        fd.shutdown_input();
-        fd.shutdown_output();
-        throw std::system_error(make_error_code(error::connection_aborted));
-      }
-      socket.emplace(std::move(fd));
-      // read server's handshake header
-      return socket->read(server_header_size);
-    }).then([this] (bufferlist headerbl) {
-      auto p = headerbl.cbegin();
-      validate_banner(p);
-      entity_addr_t saddr, caddr;
-      ::decode(saddr, p);
-      ::decode(caddr, p);
-      ceph_assert(p.end());
-      validate_peer_addr(saddr, peer_addr);
-
-      if (my_addr != caddr) {
-        // take peer's address for me, but preserve my nonce
-        caddr.nonce = my_addr.nonce;
-        my_addr = caddr;
-      }
-      // encode/send client's handshake header
-      bufferlist bl;
-      bl.append(buffer::create_static(banner_size, banner));
-      ::encode(my_addr, bl, 0);
-      h.global_seq = messenger.get_global_seq();
-      return socket->write_flush(std::move(bl));
-    }).then([=] {
-      return seastar::do_until([=] { return state == state_t::open; },
-                               [=] { return repeat_connect(); });
-    }).then([this] {
-      // start background processing of tags
-      read_tags_until_next_message();
-    }).then_wrapped([this] (auto fut) {
-      // satisfy the handshake's promise
-      fut.forward_to(std::move(h.promise));
+  seastar::with_gate(dispatch_gate, [this] {
+      return seastar::connect(peer_addr.in4_addr())
+        .then([this](seastar::connected_socket fd) {
+          if (state == state_t::closing) {
+            fd.shutdown_input();
+            fd.shutdown_output();
+            throw std::system_error(make_error_code(error::connection_aborted));
+          }
+          socket.emplace(std::move(fd));
+          // read server's handshake header
+          return socket->read(server_header_size);
+        }).then([this] (bufferlist headerbl) {
+          auto p = headerbl.cbegin();
+          validate_banner(p);
+          entity_addr_t saddr, caddr;
+          ::decode(saddr, p);
+          ::decode(caddr, p);
+          ceph_assert(p.end());
+          validate_peer_addr(saddr, peer_addr);
+
+          if (my_addr != caddr) {
+            // take peer's address for me, but preserve my nonce
+            caddr.nonce = my_addr.nonce;
+            my_addr = caddr;
+          }
+          // encode/send client's handshake header
+          bufferlist bl;
+          bl.append(buffer::create_static(banner_size, banner));
+          ::encode(my_addr, bl, 0);
+          h.global_seq = messenger.get_global_seq();
+          return socket->write_flush(std::move(bl));
+        }).then([this] {
+          return seastar::repeat([this] {
```

Comment:
What I'm doing here is to do explicit internal state changes, and at the same time, trigger related prototype logics atomically. This needs to reorganize the original ad-hoc state changes in the middle of functions, like:
https://github.com/ceph/ceph/blob/1b57df1d514588b1c8232fe0c348b21fcc90c4ee/src/crimson/net/SocketConnection.cc#L525
and:
https://github.com/ceph/ceph/blob/1b57df1d514588b1c8232fe0c348b21fcc90c4ee/src/crimson/net/SocketConnection.cc#L670

This brings us the chance to understand which operations are being executed within the exact `state_t`, so that it is much easier for us to abort the current operations cleanly, handle failure situations, and switch to other states.

And it is the foundation for the following refactorings to allow reconnect inside `Connection`, state-aware `write_message()` and `keep_alive()`, state-machine formalization, and the final goal is to abstract the entire state-aware architecture for v2 implementation.

Reply:
i see. it's all about making the state transition more visible and explicit.